cvmfs/publish/repository_transaction.cc

```diff
@@ -34,8 +34,6 @@ void Publisher::CheckTransactionStatus() {
     settings_.transaction().spool_area().publishing_lock();
   is_publishing_ =
     ServerLockFile::IsLocked(publishing_lock, false /* ignore_stale */);
-
-  session_ = new Session(settings_, llvl_);
```

Comment:
I was wondering about this.  There was nothing obvious within `Publisher::TransactionRetry()` that seemed to rely on the session object being recreated.  Is this required?

Reply:
I think so. The `session_` variable is a unique pointer.  By recreating the object, its destructor drops the lease so that the following `TransactionImpl()` call can re-acquire it.

Reply:
Thanks.  I'll update the PR, most likely by ensuring that `session_->Drop()` is called explicitly so that this intention becomes evident in the code.

Reply:
It looks as though `session_->Drop()` will already be called anyway for any exception other than `kFailTransactionState`: https://github.com/cvmfs/cvmfs/blob/f78763335b3a9c9060886d89929912c4bfd55ae4/cvmfs/publish/repository_transaction.cc#L55-L60

In the case of `kFailTransactionState`, either:
* there was no pre-existing lease file, in which case `has_lease_` would be false since the exception is raised before the call to `Acquire()` in `TransactionImpl()`
* there was a pre-existing lease file, in which case `keep_alive_` would be true

In both of these cases the `Session` destructor is a no-op, and so there is no need to reinitialise `session_`.

The overall logic here is fairly convoluted and would be improved by using RAII to automatically release the transaction lock and drop the session, rather than splitting the logic between `TransactionRetry()` and `TransactionImpl()`.  However, that cleanup can be the subject of a future PR.

For now, I am going to leave the code as-is, since reinitialising the `session_` object is unnecessary, and doing so would merely add to the confusion.